File: src/main/java/com/aleosiss/diageticui/network/NetworkService.java
package com.aleosiss.diageticui.network;

import com.aleosiss.diageticui.data.ContainerType;
import com.aleosiss.diageticui.network.handlers.MakeInventoryRequestHandler;
import com.aleosiss.diageticui.network.handlers.ReceiveInventoryRequestHandler;
import com.aleosiss.diageticui.network.packets.PacketGetInventory;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.client.MinecraftClient;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.util.math.BlockPos;

public class NetworkService {
    private static NetworkService SINGLETON;
    private NetworkService() {}

    public static NetworkService getInstance() {
        if(SINGLETON == null) {
            SINGLETON = new NetworkService();
        }

        return SINGLETON;
    }

    public void initializeServer() {
        ServerPlayNetworking.registerGlobalReceiver(NetworkConstants.DIAGETIC_INVENTORY_REQUEST_PACKET, ReceiveInventoryRequestHandler.SERVER_RECEIVE_INVENTORY_REQUEST);
    }

    @Environment(EnvType.CLIENT)
    public void initializeClient() {
        ClientPlayNetworking.registerGlobalReceiver(NetworkConstants.DIAGETIC_INVENTORY_RESPONSE_PACKET, MakeInventoryRequestHandler.CLIENT_SEND_INVENTORY_REQUEST);
    }

    public PacketByteBuf buildInventoryRequestPacket(BlockEntity blockEntity) {
        PacketByteBuf packetByteBuf = PacketByteBufs.create();
        packetByteBuf.writeBlockPos(blockEntity.getPos());

        return packetByteBuf;
    }



    //TODO: Move deserialization here
}

File: src/main/java/com/aleosiss/diageticui/network/NetworkService.kt
package com.aleosiss.diageticui.network

import com.aleosiss.diageticui.client.DiageticUIClient
import com.aleosiss.diageticui.network.NetworkConstants.DIAGETIC_INVENTORY_REQUEST_PAYLOAD
import com.aleosiss.diageticui.network.NetworkConstants.DIAGETIC_INVENTORY_RESPONSE_PAYLOAD
import com.aleosiss.diageticui.network.handlers.MakeInventoryRequestHandler
import com.aleosiss.diageticui.network.handlers.ReceiveInventoryRequestHandler
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.block.entity.BlockEntity
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.packet.CustomPayload


class NetworkService private constructor() {
    fun initializeServer() {
        val id: CustomPayload.Id<NetworkConstants.InventoryRequestPayload> = DIAGETIC_INVENTORY_REQUEST_PAYLOAD
        PayloadTypeRegistry.playC2S().register(id, NetworkConstants.InventoryRequestPayload.CODEC);
        val handler: ServerPlayNetworking.PlayPayloadHandler<NetworkConstants.InventoryRequestPayload> = ReceiveInventoryRequestHandler()
        ServerPlayNetworking.registerGlobalReceiver(id, handler)
    }

    @Environment(EnvType.CLIENT)
    fun initializeClient() {
        val id: CustomPayload.Id<NetworkConstants.InventoryResponsePayload> = NetworkConstants.DIAGETIC_INVENTORY_RESPONSE_PAYLOAD
        PayloadTypeRegistry.playS2C().register(id, NetworkConstants.InventoryResponsePayload.CODEC);
        val handler: ClientPlayNetworking.PlayPayloadHandler<NetworkConstants.InventoryResponsePayload> = MakeInventoryRequestHandler(DiageticUIClient.HUD_OVERLAY_RENDERER)
        ClientPlayNetworking.registerGlobalReceiver(id, handler)
    }

    companion object {
        private var SINGLETON: NetworkService? = null
        val instance: NetworkService
            get() {
                if (SINGLETON == null) {
                    SINGLETON = NetworkService()
                }
                return SINGLETON as NetworkService
            }
    }
}
